dmutils/repoutils/freeze-requirements.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
+"""
+Output a file containing the full, frozen dependencies of a requirements file.
+
+This will error out on the pip
+    > Requirement file x contains y, but that package is not installed...
+error caused by `pip install` being case insensitive and the `pip freeze` command being case sensitive.
+
+Usage:
+    freeze-requirements <target>
+    freeze-requirements <target> [options]
+
+Options:
+    <target>             A requirements file for us to freeze
+    --outfile=<outfile>  Output file for frozen requirement values [default: requirements.txt]
+    --verbose
+    -h, --help           Show this screen
+
+Examples:
+    python -m dmutils.repoutils.freeze-requirements requirements-app.txt
+    freeze-requirements.py requirements-app.txt --verbose
+    python -m dmutils.repoutils.freeze-requirements requirements-app.txt --outfile=not-requirements.txt --verbose
+"""
+import logging
+import re
+import shutil
+import subprocess
+import venv
+
+from docopt import docopt
+
+
+logger = logging.getLogger('freeze-requirements')
+
+
+def make_clean(virtualenv_name):
+    try:
+        logger.info(f'Removing {virtualenv_name}')
+        shutil.rmtree(virtualenv_name)
+    except FileNotFoundError:
+        pass
+
+
+def exit_failure(messages, code=1):
+    for message in messages:
+        logger.error(message.decode('utf-8'))
+    exit(code)
+
+
+if __name__ == '__main__':
+    arguments = docopt(__doc__)
+    target = arguments['<target>']
+    outfile = arguments['--outfile']
+    virtualenv_name = 'venv-freeze'
+    logging.basicConfig(level=logging.WARNING if not arguments['--verbose'] else logging.DEBUG)
+
+    make_clean(virtualenv_name)
+
+    logger.info(f'Creating virtualenv: {virtualenv_name}')
+    venv.main(args=(virtualenv_name,))
+
+    logger.info(f'Installing requirements from {target} in virtualenv: {virtualenv_name}')
+    install_cmd = subprocess.Popen(
+        [f'{virtualenv_name}/bin/pip', 'install', '-r', target],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE
+    )
+    stdout, stderr = install_cmd.communicate()
```

Comment:
I think using `subprocess.run` instead of `Popen` could simplify a lot of the code here?

Reply:
```suggestion
    install_cmd = subprocess.run(
        [f'{virtualenv_name}/bin/pip', 'install', '-r', target],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE
    )
    if install_cmd.returncode >= 1:
```

Reply:
Quality catch! I was looking at the python 2 docs for subprocess https://docs.python.org/2/library/subprocess.html and it had all these warnings about pipe.

Python 3 docs on the other hand:
https://docs.python.org/3/library/subprocess.html#subprocess.run

Looks like I can use `capture_output=True` and do away with the pipes even.

This is the subprocess method we've all been waiting for! 💯 😍

Reply:
Damn, `capture_output` added in `3.7` :(